federationapi/routing/send.go

```diff
@@ -1144,6 +1158,8 @@ func (t *txnReq) lookupEvent(ctx context.Context, roomVersion gomatrixserverlib.
 		return nil, verifySigError{event.EventID(), err}
 	}
 	h := event.Headered(roomVersion)
+	t.newEventsMutex.Lock()
 	t.newEvents[h.EventID()] = true
```

Comment:
It seems like this was the actual cause of #1584 - as workers will call `lookupEvent` and we don't guard it. Why are we adding unnecessary mutexes - `haveEventsMutex`? Why are we using RWMutexes when really we just do a ton of writes before reading again?